File: backend/taxcore/usn.py
"""Калькулятор УСН: «Доходы» (6%) и «Доходы минус расходы» (15%).

Считает налог нарастающим итогом по отчётным периодам (1 кв., полугодие, 9 мес., год),
авансовые платежи к уплате за каждый период, минимальный налог (для «доходы−расходы»)
и доплату за год.

Нюансы, которые НУЖНО держать в голове (помечаются в notes):
  • «Доходы»: налог уменьшается на страховые взносы. ИП без работников — до 100%
    (вплоть до 0); ИП с работниками — не более 50%. С 2023 фикс. взносы можно учитывать
    в периоде, на который они приходятся по сроку, даже если ещё не уплачены — здесь мы
    принимаем уже подготовленную сумму взносов к вычету (caller считает её сам).
  • «Доходы минус расходы»: взносы НЕ уменьшают налог напрямую — они входят в расходы.
    По итогам года платится не меньше минимального налога (1% от доходов).
  • Региональные пониженные ставки и торговый сбор (Москва) — передаются параметром/не
    учитываются; проверить отдельно.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from decimal import Decimal

from .models import UsnObject, round_rub, to_decimal
from .params import get_params

__all__ = ["PeriodData", "PeriodResult", "UsnYearResult", "calc_usn", "usn_quick"]


@dataclass
class PeriodData:
    """Входные данные нарастающим итогом на конец отчётного периода."""

    label: str                                          # «1 квартал» / «полугодие» / «9 месяцев» / «год»
    income_cumulative: Decimal                          # доходы нарастающим итогом
    expenses_cumulative: Decimal = Decimal("0")         # расходы нарастающим итогом (для «доходы−расходы»)
    contributions_to_deduct_cumulative: Decimal = Decimal("0")  # взносы к вычету (для «доходы»), наращ. итогом


@dataclass
class PeriodResult:
    label: str
    tax_base_cumulative: Decimal             # налоговая база нарастающим итогом
    tax_before_deduction_cumulative: Decimal # налог до вычета взносов
    deduction_cumulative: Decimal            # принятые к вычету взносы (для «доходы»)
    tax_cumulative: Decimal                  # налог нарастающим итогом (с учётом вычета/мин. налога)
    advance_due_this_period: Decimal         # к уплате за этот период (за вычетом ранее уплаченного)
    notes: list[str] = field(default_factory=list)


@dataclass
class UsnYearResult:
    usn_object: UsnObject
    rate: Decimal
    periods: list[PeriodResult]
    tax_year_computed: Decimal      # расчётный налог за год (до сравнения с минимальным)
    min_tax: Decimal                # минимальный налог (только для «доходы−расходы», иначе 0)
    tax_year_final: Decimal         # итоговый налог за год
    advances_paid_total: Decimal    # сумма авансов за периоды до годового (Q1+полугодие+9мес)
    year_payment_due: Decimal       # к доплате за год
    notes: list[str] = field(default_factory=list)


def calc_usn(
    year: int,
    usn_object: UsnObject,
    periods: list[PeriodData],
    has_employees: bool = False,
    rate: Decimal | None = None,
) -> UsnYearResult:
    if not periods:
        raise ValueError("Передайте хотя бы один отчётный период (PeriodData).")

    p = get_params(year)
    if usn_object == UsnObject.INCOME:
        eff_rate = to_decimal(rate) if rate is not None else p.usn_income_rate
    else:
        eff_rate = to_decimal(rate) if rate is not None else p.usn_income_minus_rate

    notes: list[str] = []
    if not p.verified:
        notes.append(f"Параметры {year} года не сверены (verified=False) — проверить.")

    n = len(periods)
    income_year = to_decimal(periods[-1].income_cumulative)
    min_tax = (
        round_rub(income_year * p.usn_min_tax_rate)
        if usn_object == UsnObject.INCOME_MINUS
        else Decimal("0")
    )

    results: list[PeriodResult] = []
    prev_payments = Decimal("0")

    for i, pd in enumerate(periods):
        is_last = i == n - 1
        income_cum = to_decimal(pd.income_cumulative)
        period_notes: list[str] = []

        if usn_object == UsnObject.INCOME:
            base = income_cum
            tax_before = round_rub(base * eff_rate)
            ded_avail = to_decimal(pd.contributions_to_deduct_cumulative)
            if has_employees:
                max_ded = round_rub(tax_before * Decimal("0.5"))
                applied = min(ded_avail, max_ded)
                if ded_avail > max_ded:
                    period_notes.append("Вычет ограничен 50% (есть работники).")
            else:
                applied = min(ded_avail, tax_before)
            tax_after = tax_before - applied
            if tax_after < 0:
                tax_after = Decimal("0")
        else:
            exp_cum = to_decimal(pd.expenses_cumulative)
            base = income_cum - exp_cum
            if base < 0:
                base = Decimal("0")
            tax_before = round_rub(base * eff_rate)
            applied = Decimal("0")
            tax_after = tax_before

        # Минимальный налог применяется только по итогам года (последний период).
        tax_effective = tax_after
        if is_last and usn_object == UsnObject.INCOME_MINUS and min_tax > tax_after:
            tax_effective = min_tax
            period_notes.append(
                f"Применён минимальный налог 1% = {min_tax} ₽ (больше расчётного {tax_after} ₽)."
            )

        advance = tax_effective - prev_payments
        if advance < 0:
            advance = Decimal("0")  # переплата — к уменьшению, не к доплате

        results.append(
            PeriodResult(
                label=pd.label,
                tax_base_cumulative=round_rub(base),
                tax_before_deduction_cumulative=tax_before,
                deduction_cumulative=applied,
                tax_cumulative=tax_effective,
                advance_due_this_period=advance,
                notes=period_notes,
            )
        )
        prev_payments += advance

    last = results[-1]
    tax_year_computed = last.tax_before_deduction_cumulative - last.deduction_cumulative
    if tax_year_computed < 0:
        tax_year_computed = Decimal("0")

    return UsnYearResult(
        usn_object=usn_object,
        rate=eff_rate,
        periods=results,
        tax_year_computed=tax_year_computed,
        min_tax=min_tax,
        tax_year_final=last.tax_cumulative,
        advances_paid_total=sum((r.advance_due_this_period for r in results[:-1]), Decimal("0")),
        year_payment_due=last.advance_due_this_period,
        notes=notes,
    )


def usn_quick(
    year: int,
    usn_object: UsnObject,
    income,
    expenses=0,
    contributions_to_deduct=0,
    has_employees: bool = False,
    rate: Decimal | None = None,
) -> UsnYearResult:
    """Упрощённый годовой расчёт (один период «год») — для простого экрана-калькулятора."""
    period = PeriodData(
        label="год",
        income_cumulative=to_decimal(income),
        expenses_cumulative=to_decimal(expenses),
        contributions_to_deduct_cumulative=to_decimal(contributions_to_deduct),
    )
    return calc_usn(year, usn_object, [period], has_employees=has_employees, rate=rate)

File: web/src/lib/taxcore/usn.ts
/**
 * Калькулятор УСН: «Доходы» (6%) и «Доходы минус расходы» (15%).
 *
 * Считает налог нарастающим итогом по отчётным периодам (1 кв., полугодие, 9 мес., год),
 * авансовые платежи к уплате за каждый период, минимальный налог (для «доходы−расходы»)
 * и доплату за год.
 *
 * Нюансы (помечаются в notes):
 *   • «Доходы»: налог уменьшается на страховые взносы. ИП без работников — до 100%
 *     (вплоть до 0); ИП с работниками — не более 50%. С 2023 фикс. взносы можно учитывать
 *     в периоде, на который они приходятся по сроку, даже если ещё не уплачены — здесь мы
 *     принимаем уже подготовленную сумму взносов к вычету (caller считает её сам).
 *   • «Доходы минус расходы»: взносы НЕ уменьшают налог напрямую — они входят в расходы.
 *     По итогам года платится не меньше минимального налога (1% от доходов).
 *   • Региональные пониженные ставки и торговый сбор (Москва) — передаются параметром/не
 *     учитываются; проверить отдельно.
 */

import Decimal from 'decimal.js';
import { roundRub, toDecimal, type DecimalLike } from './money.js';
import { getParams, type UsnObject } from './params.js';

/** Входные данные нарастающим итогом на конец отчётного периода. */
export interface PeriodData {
  /** «1 квартал» / «полугодие» / «9 месяцев» / «год» */
  label: string;
  /** Доходы нарастающим итогом */
  income_cumulative: DecimalLike;
  /** Расходы нарастающим итогом (для «доходы−расходы») */
  expenses_cumulative?: DecimalLike;
  /** Взносы к вычету (для «доходы»), нарастающим итогом */
  contributions_to_deduct_cumulative?: DecimalLike;
}

export interface PeriodResult {
  label: string;
  /** Налоговая база нарастающим итогом */
  tax_base_cumulative: Decimal;
  /** Налог до вычета взносов */
  tax_before_deduction_cumulative: Decimal;
  /** Принятые к вычету взносы (для «доходы») */
  deduction_cumulative: Decimal;
  /** Налог нарастающим итогом (с учётом вычета/мин. налога) */
  tax_cumulative: Decimal;
  /** К уплате за этот период (за вычетом ранее начисленного; ≥ 0) */
  advance_due_this_period: Decimal;
  /** «К уменьшению» за период, если налог просел ниже ранее начисленных авансов (≥ 0) */
  overpayment_this_period: Decimal;
  notes: string[];
}

export interface UsnYearResult {
  usn_object: UsnObject;
  rate: Decimal;
  periods: PeriodResult[];
  /** Расчётный налог за год (до сравнения с минимальным) */
  tax_year_computed: Decimal;
  /** Минимальный налог (только для «доходы−расходы», иначе 0) */
  min_tax: Decimal;
  /** Итоговый налог за год */
  tax_year_final: Decimal;
  /** Сумма авансов, начисленных за периоды до годового (Q1+полугодие+9мес) */
  advances_paid_total: Decimal;
  /** К доплате за год (≥ 0) */
  year_payment_due: Decimal;
  /** Переплата по итогам года (≥ 0) — к зачёту/возврату (ст. 78 НК РФ) */
  year_overpayment: Decimal;
  notes: string[];
}

/**
 * Полный расчёт УСН по отчётным периодам нарастающим итогом.
 *
 * @param year - налоговый год
 * @param usnObject - объект налогообложения ('income' | 'income_minus')
 * @param periods - массив периодов (1–4), данные нарастающим итогом
 * @param hasEmployees - есть ли наёмные работники (влияет на лимит вычета взносов)
 * @param rate - региональная ставка (переопределяет базовую из params)
 */
export function calcUsn(
  year: number,
  usnObject: UsnObject,
  periods: PeriodData[],
  hasEmployees = false,
  rate?: DecimalLike
): UsnYearResult {
  if (periods.length === 0) {
    throw new Error('Передайте хотя бы один отчётный период (PeriodData).');
  }

  const p = getParams(year);
  const eff_rate =
    rate !== undefined
      ? toDecimal(rate)
      : usnObject === 'income'
        ? p.usn_income_rate
        : p.usn_income_minus_rate;

  if (eff_rate.lt(0)) {
    throw new Error(`Ставка не может быть отрицательной: ${eff_rate}`);
  }

  const notes: string[] = [];
  if (!p.verified) {
    notes.push(`Параметры ${year} года не сверены (verified=false) — проверить.`);
  }

  const n = periods.length;
  // clamp income_year ≥ 0 (отрицательный доход не имеет смысла для мин. налога)
  let income_year = toDecimal(periods[n - 1].income_cumulative);
  if (income_year.lt(0)) income_year = new Decimal('0');

  // Минимальный налог считается только для «доходы−расходы» и только по итогам года.
  const min_tax =
    usnObject === 'income_minus'
      ? roundRub(income_year.times(p.usn_min_tax_rate))
      : new Decimal('0');

  const results: PeriodResult[] = [];
  let prev_payments = new Decimal('0');

  for (let i = 0; i < n; i++) {
    const pd = periods[i];
    const is_last = i === n - 1;
    const income_cum = toDecimal(pd.income_cumulative);
    const period_notes: string[] = [];

    let base: Decimal;
    let tax_before: Decimal;
    let applied: Decimal;
    let tax_after: Decimal;

    if (usnObject === 'income') {
      base = income_cum;
      if (base.lt(0)) base = new Decimal('0');
      tax_before = roundRub(base.times(eff_rate));
      const ded_avail = toDecimal(pd.contributions_to_deduct_cumulative ?? 0);

      if (hasEmployees) {
        // Потолок 50% округляем ВНИЗ — «не более 50%» (нельзя превышать).
        const max_ded = tax_before.times('0.5').toDecimalPlaces(0, Decimal.ROUND_DOWN);
        applied = Decimal.max(new Decimal('0'), Decimal.min(ded_avail, max_ded));
        if (ded_avail.gt(max_ded) && ded_avail.gt(0)) {
          period_notes.push('Вычет ограничен 50% налога (есть работники).');
        }
      } else {
        // Без работников вычет до 100% — налог может быть уменьшен до нуля.
        applied = Decimal.max(new Decimal('0'), Decimal.min(ded_avail, tax_before));
      }
      tax_after = roundRub(tax_before.minus(applied));
      if (tax_after.lt(0)) tax_after = new Decimal('0');
    } else {
      // income_minus: взносы НЕ вычитаются из налога — они учтены в расходах.
      const exp_cum = toDecimal(pd.expenses_cumulative ?? 0);
      base = income_cum.minus(exp_cum);
      if (base.lt(0)) base = new Decimal('0');
      tax_before = roundRub(base.times(eff_rate));
      applied = new Decimal('0');
      tax_after = tax_before;
    }

    // Минимальный налог применяется ТОЛЬКО по итогам года (последний период).
    // В авансовых периодах сравнение с min_tax не делается.
    let tax_effective = tax_after;
    if (is_last && usnObject === 'income_minus' && min_tax.gt(tax_after)) {
      tax_effective = min_tax;
      period_notes.push(
        `Применён минимальный налог 1% = ${min_tax} ₽ (больше расчётного ${tax_after} ₽).`
      );
    }

    const raw = tax_effective.minus(prev_payments);
    let advance: Decimal;
    let overpayment_this_period: Decimal;
    if (raw.lt(0)) {
      // Налог нарастающим итогом просел ниже ранее начисленных авансов — переплата.
      advance = new Decimal('0');
      overpayment_this_period = raw.neg();
    } else {
      advance = raw;
      overpayment_this_period = new Decimal('0');
    }

    results.push({
      label: pd.label,
      tax_base_cumulative: roundRub(base),
      tax_before_deduction_cumulative: tax_before,
      deduction_cumulative: applied,
      tax_cumulative: tax_effective,
      advance_due_this_period: advance,
      overpayment_this_period,
      notes: period_notes,
    });

    prev_payments = prev_payments.plus(advance);
  }

  const last = results[results.length - 1];

  // tax_year_computed = расчётный налог до сравнения с мин. налогом (округлённый, ≥ 0)
  let tax_year_computed = roundRub(
    last.tax_before_deduction_cumulative.minus(last.deduction_cumulative)
  );
  if (tax_year_computed.lt(0)) tax_year_computed = new Decimal('0');

  // Авансы — сумма за все периоды кроме последнего (год)
  const advances_paid_total = results
    .slice(0, -1)
    .reduce((sum, r) => sum.plus(r.advance_due_this_period), new Decimal('0'));

  const tax_year_final = last.tax_cumulative;
  const year_payment_due = last.advance_due_this_period;

  // Переплата = начислено по графику (авансы + годовой платёж) минус фактический годовой налог.
  let year_overpayment = advances_paid_total.plus(year_payment_due).minus(tax_year_final);
  if (year_overpayment.lt(0)) year_overpayment = new Decimal('0');
  if (year_overpayment.gt(0)) {
    notes.push(
      `Переплата по авансам ${year_overpayment} ₽ — положительное сальдо ЕНС, ` +
        `можно зачесть/вернуть (ст. 78 НК РФ).`
    );
  }

  return {
    usn_object: usnObject,
    rate: eff_rate,
    periods: results,
    tax_year_computed,
    min_tax,
    tax_year_final,
    advances_paid_total,
    year_payment_due,
    year_overpayment,
    notes,
  };
}

/**
 * Упрощённый годовой расчёт (один период «год») — для простого экрана-калькулятора.
 */
export function usnQuick(
  year: number,
  usnObject: UsnObject,
  income: DecimalLike,
  options: {
    expenses?: DecimalLike;
    contributionsToDeduct?: DecimalLike;
    hasEmployees?: boolean;
    rate?: DecimalLike;
  } = {}
): UsnYearResult {
  const period: PeriodData = {
    label: 'год',
    income_cumulative: toDecimal(income),
    expenses_cumulative: toDecimal(options.expenses ?? 0),
    contributions_to_deduct_cumulative: toDecimal(options.contributionsToDeduct ?? 0),
  };
  return calcUsn(year, usnObject, [period], options.hasEmployees ?? false, options.rate);
}
